cleverhans/attacks.py

```diff
@@ -1433,3 +1433,106 @@ def generate(self, x, g, **kwargs):
+                                'y_target': tf.float32}
+
+        self.structural_kwargs = ['levels', 'thermometer',
+                                  'eps', 'steps', 'attack_step',
+                                  'inv_temp', 'anneal_rate',
+                                  'clip_min', 'clip_max']
+
+        if not isinstance(self.model, Model):
+            self.model = CallableModelWrapper(self.model, 'logits')
+
+    def generate(self, x, **kwargs):
+        """
+        Return a tensor that constructs adversarial examples for the given
+        input. Generate uses tf.py_func in order to operate over tensors.
+
+        :param x: (required) A tensor with the undiscretized input images.
+        :param y: (optional) A tensor with the true labels for an untargeted
+                  attack. If None (and y_target is None) then use the
+                  original labels the classifier assigns.
+        :param y_target: (optional) A tensor with the target labels for a
+                         targeted attack.
+        :param levels: (optional) Number of levels to discretize the input to.
+        :param thermometer: (optional) Whether to use thermometer encoding
+                            to discretize.
+        :param eps: (optional) Epsilon parameter within which the attacked
+                    image must stay.
+        :param steps: (optional) The number of iterations of the attack to run.
+        :param attack_step (optional): Parameter controlling the amount
+                                       of step per iteration of the attack.
+        :param inv_temp (optional): Inverse temperature for the LSPGA attack.
+        :param anneal_rate (optional): Anneal rate for the LSPGA attack.
+        :param clip_min: (optional float) Minimum input component value
+        :param clip_max: (optional float) Maximum input component value
+        """
+        import tensorflow as tf
+        from .attacks_tf import lspga
+        from .discretization_utils import discretize_uniform
+        self.parse_params(**kwargs)
+
+        x_projected = discretize_uniform(x, levels=self.levels,
+                                         thermometer=self.thermometer)
+
+        labels, nb_classes = self.get_or_guess_labels(x_projected, kwargs)
+
+        attack = lspga(x, model=self.model,
+                       levels=self.levels,
+                       steps=self.steps, eps=self.eps,
+                       attack_step=self.attack_step,
+                       clip_min=self.clip_min,
+                       clip_max=self.clip_max,
+                       thermometer=self.thermometer,
+                       noisy_grads=self.noisy_grads,
+                       inv_temp=self.inv_temp,
+                       y=self.y,
+                       y_target=self.y_target,
+                       anneal_rate=self.anneal_rate)
+
+        return attack
+
+    def parse_params(self, y=None, y_target=None, eps=0.3, steps=40,
```

Comment:
Not sure if this is common `cleverhans` convention, but a much simpler (and arguably hacky, i.e. it should be a private method) way to write this entire function would be
```python
def parse_params(self, **kwargs):
    self.__dict__.update(kwargs)
```